### PR TITLE
fix(data_connector): decode Postgres JSON columns as Value, not String

### LINKED ISSUE #1930: [Bug]: PostgreSQL data connector panics when decoding JSON columns as Option<String>
### Bug Description

The PostgreSQL implementation in `data-connector` defines several fields as PostgreSQL `JSON`, but uses incompatible Rust types on some read/write paths.

For `responses`, `input` and `raw_response` are stored as JSON values, but `get_response()` decodes them as `Option<String>`. `tokio-postgres` rejects this conversion, and `Row::get()` turns the conversion error into a panic.

This is visible through `/v1/responses`: the initial stored request succeeds, but a follow-up request using `previous_response_id` terminates without returning an HTTP response.

The issue reproduces against a local plaintext PostgreSQL 16 instance, without TLS, PgBouncer, or a managed database service. The failing operation is in the `data-connector` retrieval path rather than worker protocol handling.

### Steps to Reproduce

1. Launch SMG with an OpenAI-compatible worker and PostgreSQL history enabled:

```text
smg launch \
  --backend openai \
  --worker-urls http://<worker>:8000 \
  --history-backend postgres \
  --postgres-db-url postgres://user:password@postgres:5432/smg
```

2. Store a response:

```bash
FIRST=$(curl -sS http://localhost:30000/v1/responses \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "<model>",
    "input": "Remember the token PG-HISTORY-TEST.",
    "store": true
  }')

ID=$(jq -e -r '.id' <<<"$FIRST")
```

3. Continue from the stored response:

```bash
curl -sS \
  -o /tmp/smg-follow-up.json \
  -w 'HTTP status: %{http_code}\n' \
  http://localhost:30000/v1/responses \
  -H 'Content-Type: application/json' \
  -d "{
    \"model\": \"<model>\",
    \"input\": \"What token did I ask you to remember?\",
    \"previous_response_id\": \"$ID\",
    \"store\": true
  }"
```

### Expected Behavior

The second request should load the stored response and return a normal 2xx response with the previous context available.

### Actual Behavior

The client receives no HTTP response:

```text
curl: (52) Empty reply from server
HTTP status: 000
```

The request task panics while decoding `responses.input`:

```text
thread 'tokio-rt-worker' panicked at
data-connector-1.0.0/src/postgres.rs:516:46:
error retrieving column input: error deserializing column 3
```

The underlying type mismatch is:

```text
WrongType {
  postgres: Json,
  rust: "core::option::Option<alloc::string::String>"
}
```

### Component

data-connector (storage)

### Routing Policy

N/A

### Connection Mode

HTTP

### Environment

- Shepherd Model Gateway: v1.7
- Bundled `data-connector` observed at runtime: 1.0.0
- PostgreSQL: 16
- Deployment: local Docker Compose on Linux
- PostgreSQL connection: direct plaintext connection
- Streaming: disabled
- Reproduction rate: consistent

### Additional Context

The runtime reproduction above uses the bundled `data-connector` 1.0.0. Separately, a source audit of the published `data-connector` 2.3.2 shows that the same issue remains in its VCS commit (`ff679739f0112553d7fe7732cc7f7d82a82ce2c6`):

- `responses.input` and `responses.raw_response` are declared as `JSON`:
  https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L830-L838
- `get_response()` reads them as `Option<String>`:
  https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L895-L919
- `conversations.metadata` has the same mismatch:
  - JSON schema: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L115-L121
  - Serialized and bound as `Option<String>`: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L158-L175
  - Read as `Option<String>`: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L242-L254
- `conversation_items.content` is also affected:
  - JSON schema: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L366-L379
  - Serialized and bound as `String`: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L421-L458
  - Read as `Option<String>`: https://github.com/lightseekorg/smg/blob/ff679739f0112553d7fe7732cc7f7d82a82ce2c6/crates/data_connector/src/postgres.rs#L764-L794

`tokio-postgres` already enables `with-serde_json-1`, so PostgreSQL JSON values should be bound and decoded consistently as `serde_json::Value` / `Option<serde_json::Value>`, with conversion to API-facing strings or maps performed after the database boundary. Existing `NULL` semantics should be preserved.

Using `try_get()` and propagating conversion errors would also prevent unexpected database values from terminating the request task.

Suggested regression coverage:

- A PostgreSQL integration test for `store_response()` → `get_response()`.
- A PostgreSQL integration test for `get_response_chain()`.
- A gateway-level test for `/v1/responses` continuation through `previous_response_id`.

A previous report of the same type mismatch in the original SGLang repository was automatically closed for inactivity without a documented fix:

https://github.com/sgl-project/sglang/issues/15861

### PR DESCRIPTION
**Pull request by XinyueZhang369**
### Problem

Closes #1930

The Postgres backend in `data-connector` declares `responses.input`,
`responses.raw_response`, `conversations.metadata`, and
`conversation_items.content` as native Postgres `JSON` columns, but
handled them as `String` on both sides of the wire:

- **Reads** used `Row::get::<Option<String>>(...)`. `tokio-postgres`'s
  typed decoder rejects a `JSON`-typed column fetched as `String`, and
  `Row::get` panics on the mismatch (`WrongType`). This killed the
  request task with no HTTP response — reproducible via `/v1/responses`
  continuation through `previous_response_id` (store a response, then
  fetch it back).
- **Writes** (`create_conversation`, `update_conversation`,
  `create_item`) serialized the JSON value to a `String` first, then
  bound that `String` as the parameter. `tokio-postgres` resolves the
  placeholder's type from the target column (`json`/`jsonb`) and
  rejects a `String` value for it at the parameter-binding layer
  (`error serializing parameter`) — this one doesn't show up from unit
  tests, only against a real server.

### Solution

Read and write these four columns as `serde_json::Value` directly,
matching what `tokio-postgres` actually decodes/encodes `JSON`/`JSONB`
as under the `with-serde_json-1` feature (already enabled). Reads use
`try_get` (not `get`) so schema drift or malformed data returns an
error instead of panicking the request task.

## Changes

- `crates/data_connector/src/postgres.rs`:
  - `build_response_from_row`, `build_item_from_row`,
    `PostgresConversationStorage::get_conversation` now `try_get` the
    JSON columns as `Option<Value>` and propagate decode errors
    instead of panicking.
  - `create_conversation`, `update_conversation`, `create_item` now
    bind `Value` directly for the JSON columns instead of a
    pre-serialized `String`.
  - `PostgresConversationStorage::parse_metadata` now operates on
    `Value` instead of routing through the Oracle/Redis text-based
    `parse_conversation_metadata` helper (Postgres already gives us a
    parsed `Value`, not text to parse).
- `crates/data_connector/tests/postgres_integration.rs` (new): opt-in
  integration tests (env-gated, `#[ignore]`) covering
  `store_response` → `get_response` (the exact issue #1930 repro),
  `get_response_chain`, and conversation/item JSON round-trips.
- `scripts/k8s-runner-resources/agentic-services.yaml`: adds a shared
  `postgres-db` Deployment + Service in the `actions-runner-system`
  namespace, mirroring the existing `oracle-db` pattern.
- `scripts/ci_agentic_svc_deps.sh`: adds a `--postgres <host>`
  reachability check, mirroring `--oracle`.
- `.github/workflows/pr-test-rust.yml`: `unit-tests` job now checks
  the Postgres shared service and runs the new integration tests
  (`--ignored --test-threads=1`) so this regression is covered by CI
  going forward, not just local unit tests.

## Test Plan

Reproduced the exact panic from #1930 locally, then verified the fix
against both a local run and the shared `postgres-db` service in the
`fra-dev` cluster (same instance CI now points at):

```
DATA_CONNECTOR_TEST_POSTGRES_URL=postgres://postgres:postgres@<host>:5432/smg_ci_test \
  cargo test -p data-connector --test postgres_integration -- --ignored --test-threads=1

running 4 tests
test conversation_item_content_round_trips_json_column ... ok
test conversation_metadata_round_trips_json_column ... ok
test get_response_chain_walks_json_columns ... ok
test store_and_get_response_round_trips_json_columns ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

Also ran, all clean:

```
cargo +nightly fmt --all -- --check          # no output
cargo clippy --workspace --all-targets -- -D warnings   # 0 warnings
cargo test --workspace --exclude smg         # 77 passed, 0 failed
cargo test -p smg                            # 20 passed, 0 failed
```

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved PostgreSQL handling for JSON conversation metadata, conversation item content, and stored responses by using native JSON decoding/binding.
  * Enhanced error handling and defaults for NULL/missing values and non-object JSON metadata, with clearer storage errors.

* **Tests**
  * Added unit coverage for metadata parsing edge-cases.
  * Added opt-in PostgreSQL integration tests for JSON round-trips, response chains, and conversation updates.
  * Updated CI to provision an isolated Postgres test database, run targeted Postgres tests with a timeout, and always clean up afterward.

* **Chores**
  * Added CI/Kubernetes Postgres support, including new helper commands, retrying port checks, and a dedicated runner Postgres deployment/service.

### CODE REVIEW

File: crates/data_connector/tests/postgres_integration.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+//! `conversation_items.content` columns are native Postgres `JSON` columns.
+//! `tokio-postgres` panics with `WrongType` if a `JSON` column is fetched as
+//! `String` instead of `serde_json::Value` — these tests exercise the real
+//! read paths (`get_response`, `get_response_chain`, `get_conversation`,
+//! `get_item`) against a real server, since that panic can't be triggered
+//! through a unit test with no database.
+//!
+//! No CI service currently provisions Postgres for this crate's `cargo test`
+//! job, so these are opt-in. Run locally with:
+//!
```
[the remark this answers is not in the document]
Comment (XinyueZhang369) [reply crates/data_connector/tests/postgres_integration.rs]: Fixed in ed48038d — updated the doc comment to reflect that this PR adds the postgres-db CI service and runs these tests with --ignored.

File: .github/workflows/pr-test-rust.yml [hunk 2]
```diff
@@ -312,6 +312,17 @@ jobs:
...
+      - name: Check Postgres shared service
+        run: bash scripts/ci_agentic_svc_deps.sh check --postgres postgres-db
+
+      - name: Run Postgres data-connector integration tests
+        timeout-minutes: 10
+        env:
+          DATA_CONNECTOR_TEST_POSTGRES_URL: postgres://postgres:postgres@postgres-db:5432/smg_ci_test
+        run: |
+          source "$HOME/.cargo/env"
+          cargo test -p data-connector --test postgres_integration -- --ignored --test-threads=1
```
[the remark this answers is not in the document]
Comment (XinyueZhang369) [reply .github/workflows/pr-test-rust.yml]: Fixed in a5f4f8f2 — each CI run now creates its own database (`create-postgres-db`/`cleanup-postgres-db` in ci_agentic_svc_deps.sh, mirroring the existing per-run Oracle user pattern) instead of sharing smg_ci_test, so concurrent workflow runs no longer race on first-time CREATE TABLE/migration DDL.

File: crates/data_connector/src/postgres.rs [hunk 3]
```diff
@@ -138,11 +135,19 @@ impl PostgresConversationStorage {
...
+        metadata: Option<Value>,
     ) -> Result<Option<ConversationMetadata>, ConversationStorageError> {
-        crate::common::parse_conversation_metadata(metadata)
-            .map_err(ConversationStorageError::StorageError)
+        match metadata {
+            None | Some(Value::Null) => Ok(None),
+            Some(Value::Object(map)) => Ok(Some(map)),
+            Some(other) => Err(ConversationStorageError::StorageError(format!(
+                "expected JSON object for conversation metadata, got: {other}"
+            ))),
```
[the remark this answers is not in the document]
Comment (XinyueZhang369) [reply crates/data_connector/src/postgres.rs:166]: Fixed in ada680ca — the non-object metadata error no longer interpolates the raw JSON value, only its type name (e.g. "expected JSON object for conversation metadata, got JSON array").

File: .github/workflows/pr-test-rust.yml [hunk 4]
```diff
@@ -306,11 +306,21 @@ jobs:
           python -m pip install transformers pillow numpy scipy
           python crates/multimodal/scripts/generate_vision_golden.py
 
+      - name: Create isolated Postgres test database
+        run: |
+          bash scripts/ci_agentic_svc_deps.sh check --postgres postgres-db
+          bash scripts/ci_agentic_svc_deps.sh create-postgres-db postgres-db
```
[the remark this answers is not in the document]
Comment (XinyueZhang369) [reply .github/workflows/pr-test-rust.yml:316]: Fixed in ada680ca — added continue-on-error: true on the DB-creation step and gated the Postgres integration test invocation on DATA_CONNECTOR_TEST_POSTGRES_URL being set, so a flaky Postgres pod no longer takes down the ~100 non-Postgres unit tests.

File: .github/workflows/pr-test-rust.yml [hunk 5]
```diff
@@ -306,11 +306,30 @@ jobs:
           python -m pip install transformers pillow numpy scipy
           python crates/multimodal/scripts/generate_vision_golden.py
 
+      - name: Create isolated Postgres test database
+        # Don't let a flaky/restarting shared Postgres service fail the
+        # whole job — the ~100 non-Postgres unit tests below don't need a
+        # database. The Postgres integration tests are skipped (not failed)
+        # if DATA_CONNECTOR_TEST_POSTGRES_URL never gets set.
+        continue-on-error: true
```
[the remark this answers is not in the document]
Comment (XinyueZhang369) [reply .github/workflows/pr-test-rust.yml]: Agreed, fixed in aa6e676f — reverted the continue-on-error/silent-skip from the previous commit. check_port now retries with a short backoff (6 attempts, 5s apart) so a mid-restart Postgres pod self-heals, but a genuinely-down service still hard-fails the job instead of letting the regression suite no-op.

**Review by slin1237 — APPROVED** [review]
